lib/vector-core/src/transform/mod.rs

```diff
@@ -109,6 +146,14 @@ pub trait FunctionTransform: Send + dyn_clone::DynClone + Sync {
 
 dyn_clone::clone_trait_object!(FunctionTransform);
 
+/// Similar to `FunctionTransform`, but with a second output for events that encountered an error
+/// during processing.
+pub trait FallibleFunctionTransform: Send + dyn_clone::DynClone + Sync {
+    fn transform(&mut self, output: &mut Vec<Event>, errors: &mut Vec<Event>, event: Event);
+}
```

Comment:
Per my comment [here](https://github.com/vectordotdev/vector/pull/9169/files#r709658133) I still do wonder if we might be able to just extend `FunctionTransform` to cover the need for error reporting. Having two almost identical traits is a touch unfortunate, unless having one that cannot fail opens up optimization paths for us at the topology level.